waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,632 @@
+;;
+;;       Copyright (c) 2017 Two Sigma Investments, LP.
+;;       All Rights Reserved
+;;
+;;       THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF
+;;       Two Sigma Investments, LP.
+;;
+;;       The copyright notice above does not evidence any
+;;       actual or intended publication of such source code.
+;;
+(ns waiter.scheduler.kubernetes
+  (:require [clj-time.core :as t]
+            [clojure.core.async :as async]
+            [clojure.data.json :as json]
+            [clojure.edn :as edn]
+            [clojure.java.io :as io]
+            [clojure.java.shell :as shell]
+            [clojure.string :as string]
+            [clojure.tools.logging :as log]
+            [plumbing.core :as pc]
+            [slingshot.slingshot :as ss]
+            [waiter.scheduler :as scheduler]
+            [waiter.service-description :as sd]
+            [waiter.util.date-utils :as du]
+            [waiter.util.http-utils :as http-utils]
+            [waiter.util.utils :as utils])
+  (:import (org.joda.time.format DateTimeFormat)))
+
+(defn authorization-from-environment []
+  "Sample implementation of the authentication string refresh function.
+   Returns a string to be used as the value for the Authorization HTTP header,
+   reading the string from the WAITER_K8S_AUTH_STRING environment variable."
+  (log/info "called waiter.scheduler.kubernetes/authorization-from-environment")
+  (System/getenv "WAITER_K8S_AUTH_STRING"))
+
+(def k8s-api-auth-str
+  "Atom containing authentication string for the Kubernetes API server.
+   This value may be periodically refreshed asynchronously."
+  (atom nil))
+
+(def k8s-timestamp-format
+  "Kubernetes reports dates in ISO8061 format, sans the milliseconds component."
+  (DateTimeFormat/forPattern "yyyy-MM-dd'T'HH:mm:ss'Z'"))
+
+(defn- timestamp-str->datetime
+  "Parse a Kubernetes API timestamp string."
+  [k8s-timestamp-str]
+  (du/str-to-date k8s-timestamp-str k8s-timestamp-format))
+
+(defn- use-short-service-hash? [k8s-max-name-length]
+  ;; This is fairly arbitrary, but if we have at least 48 characters for the app name,
+  ;; then we can fit the full 32 character service-id hash, plus a hyphen as a separator,
+  ;; and still have 25 characters left for some prefix of the app name.
+  ;; If we have fewer than 48 characters, then we'll probably want to shorten the hash.
+  (< k8s-max-name-length 48))
+
+;; Kubernetes Pods have a unique 5-character alphanumeric suffix preceded by a hyphen.
+(def ^:const pod-unique-suffix-length 5)
+
+(defn service-id->k8s-app-name [{:keys [max-name-length] :as scheduler} service-id]
```

Comment:
Should we consider using only the hash part from the `service-id` as the `k8s-app-name`; we could expose it as the `service-hash` in the descriptor?

Reply:
If we want to optimize for the lowest possible collision probabilities, that seems like a good solution. Right now I'm trying to strike a balance between keeping the output of `kubectl get pods` readable, and keeping the probability of name conflicts in k8s low.

I'd like to keep this as-is for the current PR, but I'm open to revisiting this discussion later.